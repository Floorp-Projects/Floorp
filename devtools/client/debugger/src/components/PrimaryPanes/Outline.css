/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.outline {
  overflow-y: hidden;
}

.outline > div {
  width: 100%;
  position: relative;
}

.outline-pane-info {
  padding: 0.5em;
  width: 100%;
  font-style: italic;
  text-align: center;
  user-select: none;
  font-size: 12px;
  overflow: hidden;
}

.outline-list {
  margin: 0;
  padding: 4px 0;
  position: absolute;
  top: 25px;
  bottom: 25px;
  left: 0;
  right: 0;
  list-style-type: none;
  overflow: auto;
}

.outline-list__class-list {
  margin: 0;
  padding: 0;
  list-style: none;
}

.outline-list__class-list > .outline-list__element {
  padding-inline-start: 2rem;
}

.outline-list__class-list .function-signature .function-name {
  color: var(--theme-highlight-green);
}

.outline-list .function-signature .paren {
  color: inherit;
}

.outline-list__class h2 {
  font-weight: normal;
  font-size: 1em;
  padding: 3px 0;
  padding-inline-start: 10px;
  color: var(--blue-55);
  margin: 0;
}

.outline-list__class:not(:first-child) h2 {
  margin-top: 12px;
}

.outline-list h2:hover {
  background: var(--theme-toolbar-background-hover);
}

.theme-dark .outline-list h2 {
  color: var(--theme-highlight-blue);
}

.outline-list h2 .keyword {
  color: var(--theme-highlight-red);
}

.outline-list__class h2.focused {
  background: var(--theme-selection-background);
}

.outline-list__class h2.focused,
.outline-list__class h2.focused .keyword {
  color: var(--theme-selection-color);
}

.outline-list__element {
  padding: 3px 10px 3px 10px;
  cursor: default;
  white-space: nowrap;
}

.outline-list > .outline-list__element {
  padding-inline-start: 1rem;
}

.outline-list__element-icon {
  padding-inline-end: 0.4rem;
}

.outline-list__element:hover {
  background: var(--theme-toolbar-background-hover);
}

.outline-list__element.focused {
  background: var(--theme-selection-background);
}

.outline-list__element.focused .outline-list__element-icon,
.outline-list__element.focused .function-signature * {
  color: var(--theme-selection-color);
}

.outline-footer {
  display: flex;
  box-sizing: border-box;
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  height: 25px;
  background: var(--theme-body-background);
  border-top: 1px solid var(--theme-splitter-color);
  opacity: 1;
  z-index: 1;
  user-select: none;
}

.outline-footer button {
  color: var(--theme-body-color);
}

.outline-footer button.active {
  background: var(--theme-selection-background);
  color: var(--theme-selection-color);
}
