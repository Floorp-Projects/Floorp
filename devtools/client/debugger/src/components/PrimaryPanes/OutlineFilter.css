/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.outline-filter {
  border: 1px solid var(--theme-splitter-color);
  border-top: 0px;
}

.outline-filter-input {
  height: 24px;
  width: 100%;
  background-color: var(--theme-sidebar-background);
  color: var(--theme-body-color);
  font-size: inherit;
  user-select: text;
}

.outline-filter-input.focused {
  border: 1px solid var(--theme-highlight-blue);
}

.outline-filter-input::placeholder {
  color: var(--theme-text-color-alt);
  opacity: 1;
}

.theme-dark .outline-filter-input.focused {
  border: 1px solid var(--blue-50);
}
