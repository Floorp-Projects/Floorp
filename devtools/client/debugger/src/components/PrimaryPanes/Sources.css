/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.sources-panel {
  background-color: var(--theme-sidebar-background);
  display: flex;
  flex: 1;
  flex-direction: column;
  overflow: hidden;
  position: relative;
}

.sources-panel * {
  user-select: none;
}

.sources-clear-root {
  padding: 4px 8px;
  width: 100%;
  text-align: start;
  white-space: nowrap;
  color: inherit;
  display: flex;
  border-bottom: 1px solid var(--theme-splitter-color);
}

.sources-clear-root .home {
  background-color: var(--theme-icon-dimmed-color);
}

.sources-clear-root .breadcrumb {
  width: 5px;
  margin: 0 2px 0 6px;
  vertical-align: bottom;
  background: var(--theme-text-color-alt);
}

.sources-clear-root-label {
  margin-left: 5px;
  line-height: 16px;
}

.sources-pane {
  display: flex;
  flex: 1;
  flex-direction: column;
}

.sources-list {
  flex: 1;
  display: flex;
}

.sources-list .managed-tree {
  flex: 1;
  display: flex;
}

.sources-list .managed-tree .tree {
  padding: 4px 0;
}

.sources-list .managed-tree .tree .node {
  display: flex;
  align-items: center;
  width: 100%;
  padding: 3px 8px 3px 6px;
}

.sources-list .managed-tree .tree .tree-node:not(.focused):hover {
  background: var(--theme-toolbar-background-hover);
}

.sources-list .img {
  margin-inline-end: 4px;
}

.sources-list .tree .focused .img {
  background-color: #ffffff;
}

/* Use the same width as .img.arrow */
.sources-list .tree .img.no-arrow {
  width: 10px;
  visibility: hidden;
}

.sources-list .tree .label .suffix {
  font-style: italic;
  font-size: 0.9em;
  color: var(--theme-comment);
}

.sources-list .tree .focused .label .suffix {
  color: inherit;
}

.theme-dark .source-list .node.focused {
  background-color: var(--theme-tab-toolbar-background);
}

.sources-list .tree .label {
  display: inline-block;
  line-height: 16px;
}

.no-sources-message {
  width: 100%;
  font-style: italic;
  text-align: center;
  padding: 0.5em;
  font-size: 12px;
  user-select: none;
  justify-content: center;
  align-items: center;
}

.sources-panel .outline {
  display: flex;
  height: 100%;
}

.source-outline-tabs {
  font-size: 12px;
  width: 100%;
  background: var(--theme-body-background);
  display: flex;
  user-select: none;
  box-sizing: border-box;
  height: var(--editor-header-height);
  margin: 0;
  padding: 0;
  border-bottom: 1px solid var(--theme-splitter-color);
}

.source-outline-tabs .tab {
  align-items: center;
  background-color: var(--theme-toolbar-background);
  color: var(--theme-toolbar-color);
  cursor: default;
  display: inline-flex;
  flex: 1;
  justify-content: center;
  overflow: hidden;
  padding: 4px 8px;
  position: relative;
}

.source-outline-tabs .tab::before {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 2px;
  background-color: var(--tab-line-color, transparent);
  transition: transform 250ms var(--animation-curve),
    opacity 250ms var(--animation-curve);
  opacity: 0;
  transform: scaleX(0);
}

.source-outline-tabs .tab.active {
  --tab-line-color: var(--tab-line-selected-color);
  color: var(--theme-toolbar-selected-color);
  border-bottom-color: transparent;
}

.source-outline-tabs .tab:not(.active):hover {
  --tab-line-color: var(--tab-line-hover-color);
  background-color: var(--theme-toolbar-hover);
}

.source-outline-tabs .tab:hover::before,
.source-outline-tabs .tab.active::before {
  opacity: 1;
  transform: scaleX(1);
}

.source-outline-panel {
  flex: 1;
  overflow: auto;
}

.source-outline-panel.has-root > div {
  height: 100%;
}

.source-outline-panel.has-root .thread-header {
  margin-top: 4px;
}

.sources-list .img.blackBox {
  mask-size: 13px;
  background-color: var(--theme-icon-checked-color);
}

/*
  Custom root styles
*/
.sources-pane.sources-list-custom-root {
  display: block;
  position: relative;
}

.sources-list-custom-root .sources-pane {
  display: block;
}

.sources-list-custom-root .no-sources-message {
  position: absolute;
  top: 26px;
  right: 0;
  bottom: 0;
  left: 0;
}

/* Removes start margin when a custom root is used */
.sources-list-custom-root
  .tree
  > .tree-node[data-expandable="false"][aria-level="0"] {
  padding-inline-start: 4px;
}

.sources-list .tree-node[data-expandable="false"] .tree-indent:last-of-type {
  margin-inline-end: 0;
}

.thread-header {
  margin-top: 10px;
}

.thread-header .img.worker,
.thread-header .img.file {
  margin-top: -1px;
  margin-inline-end: 4px;
  margin-inline-start: 2px;
}

.thread-header .img.file {
  margin-top: -2px;
}

.thread-header .label {
  line-height: 15px;
}
