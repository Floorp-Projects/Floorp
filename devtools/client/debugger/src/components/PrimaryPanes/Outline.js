/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import { showMenu } from "../../context-menu/menu";
import { connect } from "../../utils/connect";
import { score as fuzzaldrinScore } from "fuzzaldrin-plus";
const classnames = require("classnames");

import { containsPosition, positionAfter } from "../../utils/ast";
import { copyToTheClipboard } from "../../utils/clipboard";
import { findFunctionText } from "../../utils/function";

import actions from "../../actions";
import {
  getSelectedSourceWithContent,
  getSymbols,
  getCursorPosition,
  getContext,
} from "../../selectors";

import OutlineFilter from "./OutlineFilter";
import "./Outline.css";
import PreviewFunction from "../shared/PreviewFunction";
import { uniq, sortBy } from "lodash";

// Set higher to make the fuzzaldrin filter more specific
const FUZZALDRIN_FILTER_THRESHOLD = 15000;

/**
 * Check whether the name argument matches the fuzzy filter argument
 */
const filterOutlineItem = (name, filter) => {
  if (!filter) {
    return true;
  }

  if (filter.length === 1) {
    // when filter is a single char just check if it starts with the char
    return filter.toLowerCase() === name.toLowerCase()[0];
  }
  return fuzzaldrinScore(name, filter) > FUZZALDRIN_FILTER_THRESHOLD;
};

// Checks if an element is visible inside its parent element
function isVisible(element, parent) {
  const parentRect = parent.getBoundingClientRect();
  const elementRect = element.getBoundingClientRect();

  const parentTop = parentRect.top;
  const parentBottom = parentRect.bottom;
  const elTop = elementRect.top;
  const elBottom = elementRect.bottom;

  return parentTop < elTop && parentBottom > elBottom;
}

export class Outline extends Component {
  constructor(props) {
    super(props);
    this.focusedElRef = null;
    this.state = { filter: "", focusedItem: null };
  }

  componentDidUpdate(prevProps) {
    const { cursorPosition, symbols } = this.props;
    if (
      cursorPosition &&
      symbols &&
      cursorPosition !== prevProps.cursorPosition
    ) {
      this.setFocus(cursorPosition);
    }

    if (
      this.focusedElRef &&
      !isVisible(this.focusedElRef, this.refs.outlineList)
    ) {
      this.focusedElRef.scrollIntoView({ block: "center" });
    }
  }

  setFocus(cursorPosition) {
    const { symbols } = this.props;
    let classes = [];
    let functions = [];

    if (symbols && !symbols.loading) {
      ({ classes, functions } = symbols);
    }

    // Find items that enclose the selected location
    const enclosedItems = [...classes, ...functions].filter(
      ({ name, location }) =>
        name != "anonymous" && containsPosition(location, cursorPosition)
    );

    if (enclosedItems.length == 0) {
      return this.setState({ focusedItem: null });
    }

    // Find the closest item to the selected location to focus
    const closestItem = enclosedItems.reduce((item, closest) =>
      positionAfter(item.location, closest.location) ? item : closest
    );

    this.setState({ focusedItem: closestItem });
  }

  selectItem(selectedItem) {
    const { cx, selectedSource, selectLocation } = this.props;
    if (!selectedSource || !selectedItem) {
      return;
    }

    selectLocation(cx, {
      sourceId: selectedSource.id,
      line: selectedItem.location.start.line,
      column: selectedItem.location.start.column,
    });

    this.setState({ focusedItem: selectedItem });
  }

  onContextMenu(event, func) {
    event.stopPropagation();
    event.preventDefault();

    const { selectedSource, flashLineRange, getFunctionText } = this.props;

    if (!selectedSource) {
      return;
    }

    const sourceLine = func.location.start.line;
    const functionText = getFunctionText(sourceLine);

    const copyFunctionItem = {
      id: "node-menu-copy-function",
      label: L10N.getStr("copyFunction.label"),
      accesskey: L10N.getStr("copyFunction.accesskey"),
      disabled: !functionText,
      click: () => {
        flashLineRange({
          start: sourceLine,
          end: func.location.end.line,
          sourceId: selectedSource.id,
        });
        return copyToTheClipboard(functionText);
      },
    };
    const menuOptions = [copyFunctionItem];
    showMenu(event, menuOptions);
  }

  updateFilter = filter => {
    this.setState({ filter: filter.trim() });
  };

  renderPlaceholder() {
    const placeholderMessage = this.props.selectedSource
      ? L10N.getStr("outline.noFunctions")
      : L10N.getStr("outline.noFileSelected");

    return <div className="outline-pane-info">{placeholderMessage}</div>;
  }

  renderLoading() {
    return (
      <div className="outline-pane-info">{L10N.getStr("loadingText")}</div>
    );
  }

  renderFunction(func) {
    const { focusedItem } = this.state;
    const { name, location, parameterNames } = func;
    const isFocused = focusedItem === func;

    return (
      <li
        key={`${name}:${location.start.line}:${location.start.column}`}
        className={classnames("outline-list__element", { focused: isFocused })}
        ref={el => {
          if (isFocused) {
            this.focusedElRef = el;
          }
        }}
        onClick={() => this.selectItem(func)}
        onContextMenu={e => this.onContextMenu(e, func)}
      >
        <span className="outline-list__element-icon">Î»</span>
        <PreviewFunction func={{ name, parameterNames }} />
      </li>
    );
  }

  renderClassHeader(klass) {
    return (
      <div>
        <span className="keyword">class</span> {klass}
      </div>
    );
  }

  renderClassFunctions(klass, functions) {
    const { symbols } = this.props;

    if (!symbols || symbols.loading || klass == null || functions.length == 0) {
      return null;
    }

    const { focusedItem } = this.state;
    const classFunc = functions.find(func => func.name === klass);
    const classFunctions = functions.filter(func => func.klass === klass);
    const classInfo = symbols.classes.find(c => c.name === klass);

    const item = classFunc || classInfo;
    const isFocused = focusedItem === item;

    return (
      <li
        className="outline-list__class"
        ref={el => {
          if (isFocused) {
            this.focusedElRef = el;
          }
        }}
        key={klass}
      >
        <h2
          className={classnames("", { focused: isFocused })}
          onClick={() => this.selectItem(item)}
        >
          {classFunc
            ? this.renderFunction(classFunc)
            : this.renderClassHeader(klass)}
        </h2>
        <ul className="outline-list__class-list">
          {classFunctions.map(func => this.renderFunction(func))}
        </ul>
      </li>
    );
  }

  renderFunctions(functions) {
    const { filter } = this.state;
    let classes = uniq(functions.map(({ klass }) => klass));
    let namedFunctions = functions.filter(
      ({ name, klass }) =>
        filterOutlineItem(name, filter) && !klass && !classes.includes(name)
    );

    let classFunctions = functions.filter(
      ({ name, klass }) => filterOutlineItem(name, filter) && !!klass
    );

    if (this.props.alphabetizeOutline) {
      namedFunctions = sortBy(namedFunctions, "name");
      classes = classes.sort();
      classFunctions = sortBy(classFunctions, "name");
    }

    return (
      <ul
        ref="outlineList"
        className="outline-list devtools-monospace"
        dir="ltr"
      >
        {namedFunctions.map(func => this.renderFunction(func))}
        {classes.map(klass => this.renderClassFunctions(klass, classFunctions))}
      </ul>
    );
  }

  renderFooter() {
    return (
      <div className="outline-footer">
        <button
          onClick={this.props.onAlphabetizeClick}
          className={this.props.alphabetizeOutline ? "active" : ""}
        >
          {L10N.getStr("outline.sortLabel")}
        </button>
      </div>
    );
  }

  render() {
    const { symbols, selectedSource } = this.props;
    const { filter } = this.state;

    if (!selectedSource) {
      return this.renderPlaceholder();
    }

    if (!symbols || symbols.loading) {
      return this.renderLoading();
    }

    const symbolsToDisplay = symbols.functions.filter(
      ({ name }) => name != "anonymous"
    );

    if (symbolsToDisplay.length === 0) {
      return this.renderPlaceholder();
    }

    return (
      <div className="outline">
        <div>
          <OutlineFilter filter={filter} updateFilter={this.updateFilter} />
          {this.renderFunctions(symbolsToDisplay)}
          {this.renderFooter()}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  const selectedSource = getSelectedSourceWithContent(state);
  const symbols = selectedSource ? getSymbols(state, selectedSource) : null;

  return {
    cx: getContext(state),
    symbols,
    selectedSource: selectedSource,
    cursorPosition: getCursorPosition(state),
    getFunctionText: line => {
      if (selectedSource) {
        return findFunctionText(line, selectedSource, symbols);
      }

      return null;
    },
  };
};

export default connect(mapStateToProps, {
  selectLocation: actions.selectLocation,
  flashLineRange: actions.flashLineRange,
})(Outline);
