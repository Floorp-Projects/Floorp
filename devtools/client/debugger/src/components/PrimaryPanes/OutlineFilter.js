/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import classnames from "classnames";

import "./OutlineFilter.css";

export default class OutlineFilter extends Component {
  state = { focused: false };

  setFocus = shouldFocus => {
    this.setState({ focused: shouldFocus });
  };

  onChange = e => {
    this.props.updateFilter(e.target.value);
  };

  onKeyDown = e => {
    if (e.key === "Escape" && this.props.filter !== "") {
      // use preventDefault to override toggling the split-console which is
      // also bound to the ESC key
      e.preventDefault();
      this.props.updateFilter("");
    } else if (e.key === "Enter") {
      // We must prevent the form submission from taking any action
      // https://github.com/firefox-devtools/debugger/pull/7308
      e.preventDefault();
    }
  };

  render() {
    const { focused } = this.state;
    return (
      <div className="outline-filter">
        <form>
          <input
            className={classnames("outline-filter-input devtools-filterinput", {
              focused,
            })}
            onFocus={() => this.setFocus(true)}
            onBlur={() => this.setFocus(false)}
            placeholder={L10N.getStr("outline.placeholder")}
            value={this.props.filter}
            type="text"
            onChange={this.onChange}
            onKeyDown={this.onKeyDown}
          />
        </form>
      </div>
    );
  }
}
