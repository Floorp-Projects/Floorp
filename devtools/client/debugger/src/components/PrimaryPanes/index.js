/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import classnames from "classnames";
import { Tab, Tabs, TabList, TabPanels } from "react-aria-components/src/tabs";

import actions from "../../actions";
import {
  getActiveSearch,
  getProjectDirectoryRootName,
  getSelectedPrimaryPaneTab,
  getAllThreads,
  getContext,
} from "../../selectors";
import { features, prefs } from "../../utils/prefs";
import { connect } from "../../utils/connect";
import { formatKeyShortcut } from "../../utils/text";

import Outline from "./Outline";
import SourcesTree from "./SourcesTree";
import AccessibleImage from "../shared/AccessibleImage";

import "./Sources.css";

class PrimaryPanes extends Component {
  constructor(props) {
    super(props);

    this.state = {
      alphabetizeOutline: prefs.alphabetizeOutline,
    };
  }

  showPane = selectedPane => {
    this.props.setPrimaryPaneTab(selectedPane);
  };

  onAlphabetizeClick = () => {
    const alphabetizeOutline = !prefs.alphabetizeOutline;
    prefs.alphabetizeOutline = alphabetizeOutline;
    this.setState({ alphabetizeOutline });
  };

  onActivateTab = index => {
    if (index === 0) {
      this.showPane("sources");
    } else {
      this.showPane("outline");
    }
  };

  renderOutlineTabs() {
    if (!features.outline) {
      return;
    }

    const sources = formatKeyShortcut(L10N.getStr("sources.header"));
    const outline = formatKeyShortcut(L10N.getStr("outline.header"));
    const isSources = this.props.selectedTab === "sources";
    const isOutline = this.props.selectedTab === "outline";

    return [
      <Tab
        className={classnames("tab sources-tab", { active: isSources })}
        key="sources-tab"
      >
        {sources}
      </Tab>,
      <Tab
        className={classnames("tab outline-tab", { active: isOutline })}
        key="outline-tab"
      >
        {outline}
      </Tab>,
    ];
  }

  renderProjectRootHeader() {
    const { cx, projectRootName } = this.props;

    if (!projectRootName) {
      return null;
    }

    return (
      <div key="root" className="sources-clear-root-container">
        <button
          className="sources-clear-root"
          onClick={() => this.props.clearProjectDirectoryRoot(cx)}
          title={L10N.getStr("removeDirectoryRoot.label")}
        >
          <AccessibleImage className="home" />
          <AccessibleImage className="breadcrumb" />
          <span className="sources-clear-root-label">{projectRootName}</span>
        </button>
      </div>
    );
  }

  renderThreadSources() {
    return <SourcesTree threads={this.props.threads} />;
  }

  render() {
    const { selectedTab, projectRootName } = this.props;
    const activeIndex = selectedTab === "sources" ? 0 : 1;

    return (
      <Tabs
        activeIndex={activeIndex}
        className="sources-panel"
        onActivateTab={this.onActivateTab}
      >
        <TabList className="source-outline-tabs">
          {this.renderOutlineTabs()}
        </TabList>
        <TabPanels
          className={classnames("source-outline-panel", {
            "has-root": projectRootName,
          })}
          hasFocusableContent
        >
          <div className="threads-list">
            {this.renderProjectRootHeader()}
            {this.renderThreadSources()}
          </div>
          <Outline
            alphabetizeOutline={this.state.alphabetizeOutline}
            onAlphabetizeClick={this.onAlphabetizeClick}
          />
        </TabPanels>
      </Tabs>
    );
  }
}

const mapStateToProps = state => {
  return {
    cx: getContext(state),
    selectedTab: getSelectedPrimaryPaneTab(state),
    sourceSearchOn: getActiveSearch(state) === "source",
    threads: getAllThreads(state),
    projectRootName: getProjectDirectoryRootName(state),
  };
};

const connector = connect(mapStateToProps, {
  setPrimaryPaneTab: actions.setPrimaryPaneTab,
  setActiveSearch: actions.setActiveSearch,
  closeActiveSearch: actions.closeActiveSearch,
  clearProjectDirectoryRoot: actions.clearProjectDirectoryRoot,
});

export default connector(PrimaryPanes);
