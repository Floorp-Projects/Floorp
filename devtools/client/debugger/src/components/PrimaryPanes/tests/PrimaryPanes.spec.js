/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";

import PrimaryPanes from "..";

describe("PrimaryPanes", () => {
  it("should not render the directory root header if the directory root has not been set.", () => {
    const { component } = render();
    expect(component).toMatchSnapshot();
  });

  it("should render the directory root header if the directory root has been set.", () => {
    const { component } = render({
      projectRootName: "mdn.com",
    });
    expect(component).toMatchSnapshot();
  });

  it("should call clearProjectDirectoryRoot if .sources-clear-root has been clicked.", () => {
    const { component, props } = render({
      projectRootName: "mdn.com",
    });
    component.find(".sources-clear-root").simulate("click");
    expect(props.clearProjectDirectoryRoot).toHaveBeenCalled();
  });
});

function generateDefaults(overrides) {
  return {
    horizontal: false,
    projectRootName: "",
    sourceSearchOn: false,
    setPrimaryPaneTab: jest.fn(),
    setActiveSearch: jest.fn(),
    closeActiveSearch: jest.fn(),
    clearProjectDirectoryRoot: jest.fn(),
    threads: [],
    ...overrides,
  };
}

function render(overrides = {}) {
  const props = generateDefaults(overrides);
  const component = shallow(<PrimaryPanes.WrappedComponent {...props} />);
  const defaultState = component.state();
  const instance = component.instance();

  instance.shouldComponentUpdate = () => true;

  return { component, props, defaultState, instance };
}
