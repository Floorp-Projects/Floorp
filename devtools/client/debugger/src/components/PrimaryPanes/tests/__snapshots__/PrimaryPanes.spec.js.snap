// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryPanes should not render the directory root header if the directory root has not been set. 1`] = `
<Tabs
  activeIndex={1}
  className="sources-panel"
  onActivateTab={[Function]}
>
  <TabList
    activeIndex={0}
    className="source-outline-tabs"
    onActivateTab={[Function]}
    vertical={false}
  >
    <Tab
      active={false}
      className="tab sources-tab"
      key="sources-tab"
    >
      Sources
    </Tab>
    <Tab
      active={false}
      className="tab outline-tab"
      key="outline-tab"
    >
      Outline
    </Tab>
  </TabList>
  <TabPanels
    activeIndex={0}
    className="source-outline-panel"
    hasFocusableContent={true}
  >
    <div
      className="threads-list"
    >
      <Connect(SourcesTree)
        threads={Array []}
      />
    </div>
    <Connect(Outline)
      alphabetizeOutline={false}
      onAlphabetizeClick={[Function]}
    />
  </TabPanels>
</Tabs>
`;

exports[`PrimaryPanes should render the directory root header if the directory root has been set. 1`] = `
<Tabs
  activeIndex={1}
  className="sources-panel"
  onActivateTab={[Function]}
>
  <TabList
    activeIndex={0}
    className="source-outline-tabs"
    onActivateTab={[Function]}
    vertical={false}
  >
    <Tab
      active={false}
      className="tab sources-tab"
      key="sources-tab"
    >
      Sources
    </Tab>
    <Tab
      active={false}
      className="tab outline-tab"
      key="outline-tab"
    >
      Outline
    </Tab>
  </TabList>
  <TabPanels
    activeIndex={0}
    className="source-outline-panel has-root"
    hasFocusableContent={true}
  >
    <div
      className="threads-list"
    >
      <div
        className="sources-clear-root-container"
        key="root"
      >
        <button
          className="sources-clear-root"
          onClick={[Function]}
          title="Remove directory root"
        >
          <AccessibleImage
            className="home"
          />
          <AccessibleImage
            className="breadcrumb"
          />
          <span
            className="sources-clear-root-label"
          >
            mdn.com
          </span>
        </button>
      </div>
      <Connect(SourcesTree)
        threads={Array []}
      />
    </div>
    <Connect(Outline)
      alphabetizeOutline={false}
      onAlphabetizeClick={[Function]}
    />
  </TabPanels>
</Tabs>
`;
